wisl/lib/debugging/wislLifter.ml

```diff
@@ -129,176 +163,101 @@ struct
-                | Cmd.Call (x', _, _, _, _) when x = x' -> None
-                | _ ->
-                    let partial = FunCall { d; target_var = x } in
-                    Some (partial, StepAgain (None, None)))
-            | WStmt.While _ -> (
-                match gil_cmd with
-                | Cmd.Call _ ->
-                    let submap =
-                      match annot.expansion_kind with
-                      | NoExpansion ->
-                          failwith "Call for While has no expansion!"
-                      | Proc p -> Proc p
-                    in
-                    let partial = While { d; submap } in
-                    Some (partial, StepAgain (None, None))
-                | _ -> failwith "first cmd of While wasn't Call!")
-            | _ -> None)
-
-    let update partial exec_data =
-      let { cmd_report; _ } = exec_data in
-      let gil_cmd = CmdReport.(cmd_report.cmd) in
-      partial |> get_partial_data |> update_partial_data exec_data;
-      let annot = CmdReport.(cmd_report.annot) in
+      Finished { ids; display; unifys; errors; cmd_kind; submap }
+
+    let update annot (d : partial_data) (exec_data : exec_data) :
+        partial_cmd_result =
       let failwith s =
         DL.failwith
-          (fun () -> [ ("gil_cmd", cmd_to_yojson gil_cmd) ])
-          ("PartialCmds.update: " ^ s)
+          (fun () ->
+            [
+              ("annot", Annot.to_yojson annot);
+              ("exec_data", exec_data_to_yojson exec_data);
+              ("partial_data", partial_data_to_yojson d);
+            ])
+          ("WislLifter.PartialCmds.update: " ^ s)
       in
-      let finished ?(submap = NoSubmap) ?fun_call_name cmd_kind =
-        let { display; ids; unifys; errors } = get_partial_data partial in
-        let ids = ids |> ExtList.to_list in
-        let unifys = unifys |> ExtList.to_list in
-        let errors = errors |> ExtList.to_list in
-        Finished
-          { ids; display; unifys; errors; cmd_kind; submap; fun_call_name }
+      let end_kind =
+        match annot.stmt_kind with
+        | Multi b -> b
+        | _ -> failwith "tried to update partial with non-Multi cmd!"
       in
-      match exec_data.kind with
-      | Final -> finished Final
-      | _ -> (
-          match partial with
-          | VarAssign { target_var; _ } -> (
-              match gil_cmd with
-              | Cmd.Assignment (v, _) when v = target_var -> finished Normal
-              | _ -> StepAgain (None, None))
-          | Dispose { d } -> (
-              match gil_cmd with
-              | Cmd.LAction (_, action, _)
-                when WislLActions.(ac_from_str action = Dispose) ->
-                  if d.ids |> ExtList.length <> 2 then
-                    failwith "expected 2 ids when finishing Dispose!";
-                  finished Normal
-              | _ -> failwith "expected Dispose LAction for Dispose!")
-          | Lookup { target_var; _ } -> (
```

Comment:
Code elegance
![image](https://user-images.githubusercontent.com/8451179/204007382-2b9b5b73-ddb4-4667-a479-9eca6cca231b.png)

This exact matching on what's executed was scary, especially if we rewrite WISL to match the theory for the GillianTutorial™️, or if I want to do that with another language
But now that you actually encode everything in the AST and simply receive info at exec time, it's much nicer